go/shuffle/read.go

```diff
@@ -89,14 +91,48 @@ func NewReadBuilder(
 
 // StartReplayRead builds and starts a read of the given journal and offset range.
 func (rb *ReadBuilder) StartReplayRead(ctx context.Context, journal pb.Journal, begin, end pb.Offset) message.Iterator {
-	var r, err = rb.buildReplayRead(journal, begin, end)
-	if err != nil {
-		return message.IteratorFunc(func() (message.Envelope, error) {
-			return message.Envelope{}, err
-		})
-	}
-	rb.start(ctx, r)
-	return r
+	var r *read
+
+	return message.IteratorFunc(func() (_ message.Envelope, err error) {
+		var attempt int
+		for {
+			if r != nil {
+				// Fall through to keep using |r|.
+			} else if r, err = rb.buildReplayRead(journal, begin, end); err != nil {
+				return message.Envelope{}, err
+			} else if err = rb.start(ctx, r); err != nil {
+				return message.Envelope{}, err
+			}
+
+			if sr := r.resp.ShuffleResponse; sr != nil && r.resp.Index != len(sr.DocsJson) {
+				return r.dequeue(), nil
+			}
+
+			// We must receive from the stream.
+			if err = r.onRead(r.doRead()); err == nil && r.resp.TerminalError == "" {
+				continue
+			} else if err == io.EOF {
+				return message.Envelope{}, err
+			} else if err == nil {
+				return message.Envelope{}, errors.New(r.resp.TerminalError)
+			}
+
+			// Stream is broken, but may be retried.
```

Comment:
Is there a case where we should stop re-trying and just return the error? I feel like maybe we should at some point just show the shard status as failed instead of continuing to re-try indefinitely. But I'd need some learning on the possible error conditions at this point and error handling for failed shards before I form a strong opinion.

Reply:
In general, we continue to retry so long as we can expect that circumstance might eventually change such that we can make meaningful progress.

That's because we already have a shard assignment who's lifetime scopes how long we want to retry, and should that assignment be removed, it'll cancel the context (aborting further retries). So long as the assignment context is valid (meaning we have the assignment and continue to hold its associated etcd lease), we keep trying.

We're also separately plumbing through terminal errors (things like schema validations), which do abort processing.
For all other retry-able errors, we _do_ want to pick up processing again after eg a multi-hour network partition is healed.